src/main/scala/chisel3/util/Counter.scala

```diff
@@ -19,27 +19,51 @@ import chisel3.internal.naming.chiselName  // can't use chisel3_ version because
   *     ...
   *   }
   * }}}
-  *
-  * @param n number of counts before the counter resets (or one more than the
-  * maximum output value of the counter), need not be a power of two
   */
 @chiselName
-class Counter(val n: Int) {
-  require(n >= 0, s"Counter value must be nonnegative, got: $n")
-  val value = if (n > 1) RegInit(0.U(log2Ceil(n).W)) else 0.U
+class Counter private (r: Range) {
+  require(r.length > 0, s"Counter range cannot be empty, got: $r")
+  require(r.start >= 0 && r.end >= 0, s"Counter range must be positive, got: $r")
+
+  private lazy val delta = math.abs(r.step)
+  private lazy val width = math.max(log2Up(r.last + 1), log2Up(r.head + 1))
```

Comment:
Not a problem, but just FYI: `abs` and `max` are defined on the various integral types, so you can write e.g. `r.step.abs` or `foo max bar`.

Reply:
Thanks :smile:

I'm still learning how to write idomatic Scala.